mmaction/utils/precise_bn.py

```diff
@@ -0,0 +1,156 @@
+# Modified from https://github.com/facebookresearch/fvcore/blob/master/fvcore/nn/precise_bn.py  # noqa
+
+import logging
+import time
+
+import mmcv
+import torch
+from mmcv.parallel import MMDistributedDataParallel
+from mmcv.runner import Hook
+from mmcv.utils import print_log
+from torch.nn import GroupNorm
+from torch.nn.modules.batchnorm import _BatchNorm
+from torch.nn.modules.instancenorm import _InstanceNorm
+from torch.nn.parallel import DataParallel, DistributedDataParallel
+from torch.utils.data import DataLoader
+
+
+def is_parallel_module(module):
+    """Check if a module is a parallel module.
+
+    The following 3 modules (and their subclasses) are regarded as parallel
+    modules: DataParallel, DistributedDataParallel,
+    MMDistributedDataParallel (the deprecated version).
+
+    Args:
+        module (nn.Module): The module to be checked.
+    Returns:
+        bool: True if the input module is a parallel module.
+    """
+    parallels = (DataParallel, DistributedDataParallel,
+                 MMDistributedDataParallel)
+    if isinstance(module, parallels):
+        return True
+    else:
+        return False
+
+
+@torch.no_grad()
+def update_bn_stats(model, data_loader, num_iters=200, logger=None):
+    """Recompute and update the batch norm stats to make them more precise.
+
+    During
+    training both BN stats and the weight are changing after every iteration,
+    so the running average can not precisely reflect the actual stats of the
+    current model.
+    In this function, the BN stats are recomputed with fixed weights, to make
+    the running average more precise. Specifically, it computes the true
+    average of per-batch mean/variance instead of the running average.
+
+    Args:
+        model (nn.Module): The model whose bn stats will be recomputed.
+        data_loader (iterator): The DataLoader iterator.
+        num_iters (int): number of iterations to compute the stats.
+        logger (:obj:`logging.Logger` | None): Logger for logging.
+            Default: None.
+    """
+
+    model.train()
+
+    assert len(data_loader) >= num_iters, (
```

Comment:
Since you set the attribute `interval` for PreciseBNHook, can we change the condition to 
`assert interval * len(data_loader) >= num_iters`?

Reply:
I'm not sure. `interval` means updating bn stats every `interval` epochs. While in this assertion, `num_iters` and `len(data_loader)` ensures for EACH epoch the steps is valid.

Do you mean by setting a total number of iters for precise bn calculation? Like if I want to do it every 5 epochs for 50 epochs. The total iters of precise bn should be (50 // 5) * len(data_loader).

Reply:
I think this way is more straight-forward. And it is consistent with the original implementation of fvcore.

Reply:
Actually, I mean if one wants to do update_bn_stats every 5 epochs, and each epoch has 60 batches. Then num_iters can be at most 300 instead of 60. Maybe useful for some small datasets.

Reply:
300 iterations will have the same effect as 60 iterations, as other parameters in the models are fixed except bn stats. 

It follows the original setting https://github.com/facebookresearch/fvcore/blob/1c6e25b1dd4d9de8af7168faabbcedb83e629c8e/fvcore/nn/precise_bn.py#L91